nvflare/app_common/ccwf/client_ctl.py

```diff
@@ -0,0 +1,630 @@
+                return make_reply(ReturnCode.BAD_REQUEST_DATA)
+            self.log_info(fl_ctx, f"got my workflow id {my_wf_id}")
+            self.workflow_id = my_wf_id
+
+            reply = self.process_config(fl_ctx)
+
+            self.engine.register_aux_message_handler(
+                topic=topic_for_end_workflow(my_wf_id),
+                message_handle_func=self._process_end_workflow,
+            )
+
+            if not reply:
+                reply = make_reply(ReturnCode.OK)
+            return reply
+
+        elif task_name == self.start_task_name:
+            self.is_starting_client = True
+
+            learnable = self.persistor.load(fl_ctx)
+            fl_ctx.set_prop(AppConstants.GLOBAL_MODEL, learnable, private=True, sticky=True)
+            initial_model = self.shareable_generator.learnable_to_shareable(learnable, fl_ctx)
+            return self.start_workflow(initial_model, fl_ctx, abort_signal)
+
+        elif task_name == self.do_learn_task_name:
+            return self._process_learn_request(shareable, fl_ctx)
+
+        elif task_name == self.report_final_result_task_name:
+            return self._process_final_result(shareable, fl_ctx)
+
+        else:
+            self.log_error(fl_ctx, f"Could not handle task: {task_name}")
+            return make_reply(ReturnCode.TASK_UNKNOWN)
+
+    @abstractmethod
+    def start_workflow(self, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
+        """
+        This is called for the subclass to start the workflow.
+        This only happens on the starting_client.
+
+        Args:
+            shareable: the initial task data (e.g. initial model weights)
+            fl_ctx: FL context
+            abort_signal: abort signal for task execution
+
+        Returns:
+
+        """
+        pass
+
+    def _get_status_report(self):
+        with self.status_lock:
+            status = self.current_status
+            must_report = False
+            if status.error:
+                must_report = True
+            elif status.timestamp:
+                must_report = True
+
+            if not must_report:
+                return None
+
+            # do status report
+            report = copy.copy(status)
+            return report
+
+    def _abort_current_task(self, fl_ctx: FLContext):
+        current_task = self.learn_task
+        if not current_task:
+            return
+
+        current_task.abort_signal.trigger(True)
+        fl_ctx.set_prop(FLContextKey.TASK_NAME, current_task.task_name)
+        self.fire_event(EventType.ABORT_TASK, fl_ctx)
+
+    def set_learn_task(self, task_data: Shareable, fl_ctx: FLContext) -> bool:
+        task_data.set_header(AppConstants.NUM_ROUNDS, self.get_config_prop(AppConstants.NUM_ROUNDS))
+        task = _LearnTask(self.learn_task_name, task_data, fl_ctx)
+        current_task = self.learn_task
+        if not current_task:
+            self.learn_task = task
+            return True
```

Comment:
Could there be multiple threads setting the learn task?

Reply:
Yes, I probably should use a lock here.

Reply:
Changed to put it inside a lock.